tooling/debugger/src/context.rs

```diff
@@ -142,6 +185,47 @@ impl<'a, B: BlackBoxFunctionSolver> DebugContext<'a, B> {
         }
     }
 
+    fn breakpoint_reached(&self) -> bool {
+        if let Some(location) = self.get_current_opcode_location() {
+            self.breakpoints.contains(&location)
+        } else {
+            false
+        }
+    }
+
+    pub(super) fn is_valid_location(&self, location: &OpcodeLocation) -> bool {
+        let opcodes = self.get_opcodes();
+        match *location {
+            OpcodeLocation::Acir(acir_index) => acir_index < opcodes.len(),
+            OpcodeLocation::Brillig { acir_index, brillig_index } => {
+                acir_index < opcodes.len()
+                    && matches!(opcodes[acir_index], Opcode::Brillig(..))
+                    && {
+                        let Opcode::Brillig(ref brillig) = opcodes[acir_index] else {
+                            unreachable!("opcode at {acir_index} is not Brillig")
+                        };
+                        brillig_index < brillig.bytecode.len()
+                    }
+            }
+        }
+    }
+
+    pub(super) fn is_breakpoint_set(&self, location: &OpcodeLocation) -> bool {
+        self.breakpoints.contains(location)
+    }
+
+    pub(super) fn add_breakpoint(&mut self, location: OpcodeLocation) {
+        _ = self.breakpoints.insert(location);
+    }
+
+    pub(super) fn delete_breakpoint(&mut self, location: &OpcodeLocation) {
+        _ = self.breakpoints.remove(location);
+    }
+
+    pub(super) fn iterate_breakpoints(&self) -> Iter<'_, OpcodeLocation> {
+        self.breakpoints.iter()
+    }
```

Comment:
We could just implement the `Iter` trait ourselves here.

Reply:
You mean so we can iterate over the debugger context directly? I'm not sure it's a good idea. We may have other collections in the context in the future and it would become ambiguous.

Reply:
Ah, that's true. Let's leave this as is.